sphinx/domains/c/_symbol.py

```diff
@@ -155,36 +162,49 @@ def _add_function_params(self) -> None:
     def remove(self) -> None:
         if self.parent is None:
             return
-        assert self in self.parent._children
-        self.parent._children.remove(self)
+        name = str(self.ident)
+        assert name in self.parent._childrenByName
+        del self.parent._childrenByName[name]
+        if self.docname in self.parent._childrenByDocname:
+            del self.parent._childrenByDocname[self.docname]
+        if self.ident.is_anon():
+            self.parent._anonChildren.remove(self)
         self.parent = None
 
     def clear_doc(self, docname: str) -> None:
-        for sChild in self._children:
-            sChild.clear_doc(docname)
-            if sChild.declaration and sChild.docname == docname:
-                sChild.declaration = None
-                sChild.docname = None
-                sChild.line = None
-                if sChild.siblingAbove is not None:
-                    sChild.siblingAbove.siblingBelow = sChild.siblingBelow
-                if sChild.siblingBelow is not None:
-                    sChild.siblingBelow.siblingAbove = sChild.siblingAbove
-                sChild.siblingAbove = None
-                sChild.siblingBelow = None
+        if docname not in self._childrenByDocname:
+            for child in self._childrenByName.values():
+                child.clear_doc(docname)
+            return
+        for sChild in self._childrenByDocname[docname]:
+            sChild.declaration = None
+            sChild.docname = None
+            sChild.line = None
+            if sChild.siblingAbove is not None:
+                sChild.siblingAbove.siblingBelow = sChild.siblingBelow
+            if sChild.siblingBelow is not None:
+                sChild.siblingBelow.siblingAbove = sChild.siblingAbove
+            sChild.siblingAbove = None
+            sChild.siblingBelow = None
+            name = str(sChild.ident)
+            if name in self._childrenByName:
+                del self._childrenByName[name]
+                if sChild.ident.is_anon():
+                    self._anonChildren.remove(sChild)
+        del self._childrenByDocname[docname]
 
     def get_all_symbols(self) -> Iterator[Symbol]:
         yield self
-        for sChild in self._children:
+        for sChild in self._childrenByName.values():
```

Comment:
```suggestion
        for sChild in self._children:
```

(After you have created `self._children = self._childrenByName.values()`)